### DISCUSSION TITLE
Character virtual, JPH debug assert triggered at update

### DISCUSSION THREAD

**Discussion by weiquanyeo** [jrouwe/JoltPhysics#893] [Q&A]
Hi, we are new to working with physics in games, recently while trying to implement character virtual for our game engine, we faced an issue with the update function where the JPH_ASSERT(mGroundState != EGroundState::InAir); throws a breakpoint in MoveToContact() function. 
May I inquire what could be the possible reason this Assert is triggering?

**Comment by jrouwe** [comment]
MoveToContact is used by WalkStairs and StickToGround to move the character to a contact that was found earlier. After this function finishes the character should be on that contact point, so should not be in the 'in air' state. I've never seen that assert trigger before and without further information I cannot really say what may be going wrong.

If you could give me:
* The callstack when the assert triggers
* The value of inContact with all of its members
* The value of mActiveContacts, all entries and for each entry all of the member variables

**Reply by weiquanyeo** [reply]
The following is the callstack  when the assert triggers:
![image](https://github.com/jrouwe/JoltPhysics/assets/90823005/fe961439-680d-45fe-9695-00ad728dec2e)
This is the mActiveContacts, there's only one contact:
![image](https://github.com/jrouwe/JoltPhysics/assets/90823005/3729cc7c-e4a7-4420-9fc1-6189493e9ebe)

As for inContact I am unable to retrieve it as we are running the library as an external lib so source file information is not being loaded.

Some additional information, I can provide regarding this is that it seems to happen more consistently when there is a smaller body between the character and the floor but it also occasionally happens when moving towards a body.

**Reply by jrouwe** [reply]
That callstack doesn't make any sense: GetFirstContactForSweep doesn't call MoveToContact and neither does MoveShape or Update.

Also if I look at the contents of mActiveContacts then mHadCollision = true and mDistance < FLT_MAX so it should have gotten to this line:

https://github.com/jrouwe/JoltPhysics/blob/cc49e07d7c4eb854a60cb21ea0abdfcc93511410/Jolt/Physics/Character/CharacterVirtual.cpp#L791

which means that 'best_contact' should be non-null:

https://github.com/jrouwe/JoltPhysics/blob/cc49e07d7c4eb854a60cb21ea0abdfcc93511410/Jolt/Physics/Character/CharacterVirtual.cpp#L847

which means that `mGroundState` should not be InAir:

https://github.com/jrouwe/JoltPhysics/blob/cc49e07d7c4eb854a60cb21ea0abdfcc93511410/Jolt/Physics/Character/CharacterVirtual.cpp#L925

Maybe you could take a look at the disassembly of GetFirstContactForSweep around the return address of the call to MoveToContact to see how it gets to MoveToContact. Perhaps it's calling a virtual function that ends up in the wrong place? Or perhaps there is some code gen issue?

Are you using a recent version of Jolt?
